File: assets/centralities.R
library(data.table)
library(igraph)
library(bit64)
library(rjson)

refs = fread('assets/data/references.csv')
refs = unique(refs)

if("weight" %in% colnames(refs)){
  refs[refs$weigh <= 0,]$weight = 0.000001  
}

G = graph_from_data_frame(refs, directed = TRUE)

## Computing centralities for each node
nodes = data.table()
nodes$paper_id = V(G)$name
nodes$betweenness = betweenness(G)
nodes$indegree = degree(G, mode = "in")
nodes$clustering = transitivity(G, 'local')
nodes$pagerank = page_rank(G)$vector



fn <- fromJSON(file = 'assets/data/fake.json')
ids = c()
for(f in fn){
  ids = c(ids, f$paper_id)
}
fn = data.table(paper_id = ids)
fn$paper_id <- as.character(fn$paper_id)

nodes = merge(fn, nodes, by = 'paper_id', all.x = TRUE)


nodes[is.na(nodes$betweenness),]$betweenness <- 0
nodes[is.na(nodes$indegree),]$indegree <- 0
nodes[is.na(nodes$clustering),]$clustering <- 0
nodes[is.na(nodes$pagerank),]$pagerank <- 0


fwrite(nodes, 'assets/data/paper_centrality.csv')

File: assets/centralities.py
import json
import pandas as pd
import os

os.system('Rscript assets/centralities.R >assets/rlog/a.Rout 2>assets/rlog/a.Rerr')

with open('assets/data/fake.json', 'r') as fr:
        fake = json.load(fr)

L0 = len(fake)

pc = pd.read_csv('assets/data/paper_centrality.csv')

fake = pd.DataFrame(fake, columns = fake[0].keys())
L1 = fake.shape[0]

fake = pd.merge(fake, pc, on = 'paper_id')
L2 = fake.shape[0]


if not L0 == L1 == L2:
    print('Lost {} papers in the process!'.format(L0 - L2))

fake = fake.to_dict(orient = 'row')

with open('assets/data/fake.json', 'w') as fw:
    json.dump(fake, fw)
